chanmqttproxy/channelsmqttproxy.py

```diff
@@ -5,24 +5,76 @@
 import os
 import signal
 import socket
+import ssl
 
 from gmqtt import Client as MQTTClient
+from gmqtt.mqtt.handler import MQTTConnectError
 from gmqtt.mqtt.constants import MQTTv311, MQTTv50
 
 LOGGER = logging.getLogger(__name__)
 
 
 class ChannelsMQTTProxy:
+    @staticmethod
+    def strtobool(val):
+        """
+        FROM: https://stackoverflow.com/a/18472142
+        Convert a string representation of truth to true (1) or false (0).
+        True values are 'y', 'yes', 't', 'true', 'on', and '1'; false values
+        are 'n', 'no', 'f', 'false', 'off', and '0'.  Raises ValueError if
+        'val' is anything else.
+        """
+        val = val.lower()
+        if val in ('y', 'yes', 't', 'true', 'on', '1'):
+            return True
+        elif val in ('n', 'no', 'f', 'false', 'off', '0'):
+            return False
+        else:
+            raise ValueError("invalid truth value %r" % (val,))
+
     def __init__(self, channel_layer, settings):
         self.channel_layer = channel_layer
 
         # MQTTClient takes an identifier which is seen at the broker
         # Creating the client does not connect.
         self.mqtt = MQTTClient(
             f"ChannelsMQTTProxy@{socket.gethostname()}.{os.getpid()}")
-        self.mqtt.set_auth_credentials(username=settings.MQTT_USER,
-                                       password=settings.MQTT_PASSWORD)
+        try:
+            self.mqtt.set_auth_credentials(username=settings.MQTT_USER,
+                                        password=settings.MQTT_PASSWORD)
+        except AttributeError:
+            # Settings are not defined. Try anonymous connection
+            pass
         self.mqtt_host = settings.MQTT_HOST
+        try:
+            self.mqtt_port = int(settings.MQTT_PORT)
+        except AttributeError:
+            # Setting not defined. Use default unsecured.
+            self.mqtt_port = 1883
+        # Set ssl
+        try:
+            self.mqtt_usessl = self.strtobool(settings.MQTT_USE_SSL)
+        except AttributeError:
+            # Setting is not defined. Assume false
+            self.mqtt_usessl = False
+
+        try:
+            self.mqtt_ssl_ca = settings.MQTT_SSL_CA
+            self.mqtt_ssl_cert = settings.MQTT_SSL_CERT
+            self.mqtt_ssl_key = settings.MQTT_SSL_KEY
+            try:
+                self.mqtt_ssl_verify = \
+                    self.strtobool(settings.MQTT_SSL_VERIFY)
```

Comment:
remove strtobool()

Reply:
There is a valid use case and a story behind this...
### The use case
`.env` file
```python
# deployment specific secrets and settings
# --- snip ---

# mqtt settings
# host, username, password, etc
# including...
MQTT_PORT=8883
MQTT_USE_SSL=True
MQTT_SSL_VERIFY=True
```

My django `settings.py` file
```python
# other includes
# --- snip ---

from dotenv import load_dotenv
load_dotenv()
env = os.environ

# --- snip ---
# standard django settings.py stuff
# --- snip ---

MQTT_PORT=env.get('MQTT_PORT', None)
MQTT_USE_SSL=env.get('MQTT_USE_SSL', None)
MQTT_SSL_VERIFY=env.get('MQTT_SSL_VERIFY', None)
```

By the time they reach `channelsmqttproxy.py`:
```
type(settings.MQTT_PORT) = <class 'str'>
type(settings.MQTT_USE_SSL) = <class 'str'>
type(settings.MQTT_SSL_VERIFY) = <class 'str'>
```
### The story
Now, the obvious stuff
```python
int(2) = 2
int('2') = 2
bool(True) = True
bool('True') = True
bool(False) = False
```
but, here's the thing that got me...
```python
bool('False') = True # Makes sense when you stop and think about it but doesn't seem intuitive for this case
```
### Proposal
Above these blocks with the `strtobool()` calls, there is already a call to `int()` to ensure `MQTT_PORT` is an int or can be converted to an int.
Given that there is already protection for `MQTT_PORT`, I would like to include the same for `MQTT_USE_SSL` and `MQTT_SSL_VERIFY`.

Having said all that, I would like to
* Rename `strtobool()` to `tobool()` to be closer to its intent.
* Have `tobool()` behave in the same manner as `int()` would i.e.
```python 
tobool(True) = True
tobool('True') = True
tobool(False) = False
tobool('False') = False # Fixed for intent
```

Please let me know if you are happy with that and I will proceed with the changes.

Reply:
Oh, and I'll remember to run it through the linter this time too.

Reply:
I see the issue - however I think the settings.py should be documented as taking bool.
The solution would be to have this in settings.py:
```
   from mysettings.helper import tobool
   
    MQTT_USE_SSL = tobool(env.get('MQTT_USE_SSL', None))
````

Doesn't that make more sense?